File: algorithms/[190]颠倒二进制位/solution.ts
// topics = ["位运算"]

function reverseBits(n: number): number {
  let res = 0;

  // JavaScript 中参与位运算的值会先被转为 32 位有符号整型！
  for (let i = 0; i < 32; i += 1) {
    let t = (n >> i) & 1;
    if (t) {
      res |= t << (31 - i);
    }
  }

  // 逻辑右移/无符号右移，最高位补 0
  return res >>> 0;
}

File: algorithms/[190]颠倒二进制位/solution.py
# topics = ["位运算"]


class Solution:
    def reverseBits(self, n: int) -> int:
        res = 0

        for i in range(32):
            # 读取二进制的第 i 位
            t = (n >> i) & 1
            if t:
                # 设置二进制的对称位
                res |= t << (31 - i)

        return res
